File: Leetcode/152-Maximum-Product-Subarray/MaximumProductSubarray.java
public class MaximumProductSubarray {
    public int maxProduct(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        
        int maxProduct = nums[0];
        int maxTemp = nums[0];
        int minTemp = nums[0];
        
        for (int i=1; i<nums.length; i++) {
            int a = maxTemp*nums[i];
            int b = minTemp*nums[i];
            maxTemp = Math.max(Math.max(a,b), nums[i]);
            minTemp = Math.min(Math.min(a,b), nums[i]);
            maxProduct = Math.max(maxProduct, maxTemp);
        }
        return maxProduct; 
    }
}

File: Leetcode/152-Maximum-Product-Subarray/MaximumProductSubarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    if (!nums && nums.length==0) {
        return 0
    }
    var maxProduct = nums[0]
    var maxTemp = nums[0]
    var minTemp = nums[0]

    for (var i=1; i<nums.length; i++) {
        var a = maxTemp*nums[i]
        var b = minTemp*nums[i]
        maxTemp = Math.max(Math.max(a,b), nums[i])
        minTemp = Math.min(Math.min(a,b), nums[i])
        maxProduct = Math.max(maxProduct, maxTemp)
    }
    return maxProduct
};
